api/repositories/role_repository.go

```diff
@@ -0,0 +1,149 @@
+package repositories
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"time"
+
+	"code.cloudfoundry.org/cf-k8s-controllers/api/repositories/authorization"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	hnsv1alpha2 "sigs.k8s.io/hierarchical-namespaces/api/v1alpha2"
+)
+
+var (
+	ErrorDuplicateRoleBinding          = errors.New("RoleBinding with that name already exists")
+	ErrorMissingRoleBindingInParentOrg = errors.New("no RoleBinding found in parent org")
+)
+
+const (
+	RoleGuidLabel         = "cloudfoundry.org/role-guid"
+	RoleUserLabel         = "cloudfoundry.org/role-user"
+	RoleTypeLabel         = "cloudfoundry.org/role-type"
+	roleBindingNamePrefix = "cf"
+)
+
+//counterfeiter:generate -o fake -fake-name AuthorizedInChecker . AuthorizedInChecker
+
+type AuthorizedInChecker interface {
+	AuthorizedIn(ctx context.Context, identity authorization.Identity, namespace string) (bool, error)
+}
+
+//+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=rolebindings,verbs=create
+
+type RoleRecord struct {
+	GUID      string
+	CreatedAt time.Time
+	UpdatedAt time.Time
+	Type      string
+	Space     string
+	User      string
+}
+
+type RoleRepo struct {
+	privilegedClient    client.Client
+	roleMappings        map[string]string
+	authorizedInChecker AuthorizedInChecker
+}
+
+func NewRoleRepo(privilegedClient client.Client, authorizedInChecker AuthorizedInChecker, roleMappings map[string]string) *RoleRepo {
+	return &RoleRepo{
+		privilegedClient:    privilegedClient,
+		roleMappings:        roleMappings,
+		authorizedInChecker: authorizedInChecker,
+	}
+}
+
+func (r *RoleRepo) CreateSpaceRole(ctx context.Context, role RoleRecord) (RoleRecord, error) {
+	k8sRoleName, ok := r.roleMappings[role.Type]
+	if !ok {
+		return RoleRecord{}, fmt.Errorf("invalid role type: %q", role.Type)
+	}
+
+	orgName, err := r.getOrgName(ctx, role.Space)
+	if err != nil {
+		return RoleRecord{}, err
+	}
+
+	userIdentity := authorization.Identity{
+		Name: role.User,
+		Kind: rbacv1.UserKind,
+	}
+
+	hasOrgBinding, err := r.authorizedInChecker.AuthorizedIn(ctx, userIdentity, orgName)
+	if err != nil {
+		return RoleRecord{}, fmt.Errorf("failed to check for role in parent org: %w", err)
+	}
+
+	if !hasOrgBinding {
+		return RoleRecord{}, ErrorMissingRoleBindingInParentOrg
+	}
+
+	roleBinding := rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: role.Space,
+			Name:      calculateRoleBindingName(role),
```

Comment:
It makes sense encoding subject and role name in the rolebinding name here. Although breaking the convention of using a guid in the name, it makes sense to have a deterministic name here since this name is immutable in CF. Hence we protect from duplicate rolebindings using built-in k8s name unqiueness checking.